app/commands/user_track/generate_concept_exercise_mapping.rb

```diff
@@ -5,30 +5,43 @@ class GenerateConceptExerciseMapping
     initialize_with :user_track
 
     def call
-      concept_exercise_mapping =
-        user_track.
-          track.
-          exercises.
-          # includes(:taught_concepts, :prerequisites). <-- How to optimize? raises error d/t practice exercise
-          each_with_object({}) do |exercise, hash|
-            concepts = exercise.practice_exercise? ? exercise.prerequisites : exercise.taught_concepts
+      return {} if user_track.nil?
 
-            concepts.to_a.each do |concept|
-              slug = concept.slug
-              hash[slug] ||= { exercises: 0, exercises_completed: 0 }
-              hash[slug][:exercises] += 1
-            end
-          end
+      taught_counts =
+        Exercise::TaughtConcept.
+          where(exercise_id: user_track.track.concept_exercises).
+          group(:track_concept_id).
+          count
+      taught_counts.default = 0
 
-      concept_exercise_mapping.tap do |hash|
-        user_track.solutions.completed.each do |solution|
-          exercise = solution.exercise
-          concepts = exercise.practice_exercise? ? exercise.prerequisites : exercise.taught_concepts
-          concepts.to_a.each do |concept|
-            slug = concept.slug
-            hash[slug][:exercises_completed] += 1
-          end
-        end
+      prereq_counts =
+        Exercise::Prerequisite.
+          where(exercise_id: user_track.track.practice_exercises).
+          group(:track_concept_id).
+          count
+      prereq_counts.default = 0
+
+      solved_taught_counts =
+        Exercise::TaughtConcept.
+          where(exercise_id:
+            Solution.where(exercise_id: user_track.track.concept_exercises).completed.select(:exercise_id)).
+          group(:track_concept_id).
+          count
+      solved_taught_counts.default = 0
+
+      solved_prereq_counts =
+        Exercise::Prerequisite.
+          where(exercise_id:
+            Solution.where(exercise_id: user_track.track.practice_exercises).completed.select(:exercise_id)).
+          group(:track_concept_id).
+          count
+      solved_prereq_counts.default = 0
+
+      user_track.track.concepts.each_with_object({}) do |concept, hash|
+        hash[concept.slug] = {
+          exercises: prereq_counts[concept.id] + taught_counts[concept.id],
+          exercises_completed: solved_prereq_counts[concept.id] + solved_taught_counts[concept.id]
```

Comment:
If you use `.to_i` here, then you can remove the `default = 0` lines above. Makes it a bit more concise.

```suggestion
          exercises: prereq_counts[concept.id.to_i] + taught_counts[concept.id.to_i],
          exercises_completed: solved_prereq_counts[concept.id.to_i] + solved_taught_counts[concept.id.to_i]
```

Reply:
done